### ISSUE TITLE
9.3.2.2  Keine unerwartete Kontextänderung bei Eingabe

### ISSUE THREAD

**Issue by MarcHaunschild** [BIK-BITV/BIK-Web-Test#232]
Ich fände es gut, den Begriff "Kontext" zu klären. Es ist mehrfach davon die Rede, dass Inhalte durch Eingabe (z.B. Setzen eines Check-Box-Häkchens) geändert werden. Das steht so auch in der WCAG (wobei das für mich immer mehr ein Schande of content ist, aber geschenkt).

Was aber auch in der WCAG steht und hier in der Prüfschritt-Beschreibung meiner Meinung nach fehlt ist "Opening a new window, moving focus to a different component, going to a new page (including anything that would look to a user as if they had moved to a new page) or significantly re-arranging the content of a page are examples of changes of context." - in meinen Tests finde ich häufig den Fall, dass beim Setzen von Filtern vor langen Listen der Fokus an den Seitenanfang springt. Meiner Meinung nach kann das den Eindruck erwecken, als wäre man auf einer neuen Seite und der Fokus wurde auf eine andere Komponente (üblicherweise den header) gesetzt.

Dann hätte man entsprechend eine negative Bewertung zu vergeben.

**Comment by detlevhfischer** [comment]
Wenn der Fokus versetzt wird, etwa an den Seitenanfang, ist das ein Punkt, den ich eher bei Fokus-Management in Prüfschritt 9.2.4.3 (Fokusreihenfolge) unterbringen bzw. negativ bewerten würde. OK, im Understandig Text steht "moving focus to a different component" - ich habe das bislang nicht als bloße Verschiebung des Fokus auf ein anders Element des gleichen Seiteninhalts verstanden. Anders wäre es (konstruierte Beispiel), wenn ich bei Texteingabe einen Popup (einen "different component") öffne und da den Fokus reinsetze.

**Comment by MarcHaunschild** [comment]
Natürlich kann ich mich irren (passiert ständig), aber ich meine es sprechen doch einige Gründe dafür, hier ein FAIL zu vergeben:
1. Es ist ein Prüfschritt über Vorhersagbarkeit (Make Web pages appear and operate in predictable ways.) Ich finde es nicht vorhersehbar, dass man an irgendeine Stelle versetzt wird, wenn man eine checkbox markiert
2. Moving focus to a different component in der WCAG verstehe ich, dass man die Komponente, mit der man agiert (ein fieldset oder gar das gesamte Formular) verlässt und an eine andere Stelle gelangt. Eine Komponente definiert die WCAG: a part of the content that is perceived by users as a single control for a distinct function - ein Formular, ein fieldset, ja sogar eine einzelne Checkbox klingt für mich nach "Single control for a distinct function" - bis zum Seitenanfang ist es ein weiter Weg
3. Sprung auf eine andere Seite oder der Eindruck, man sei auf einer anderen Seite - ein page reload (was in der Regel der Grund dafür ist, dass man am Seitenanfang landet) fühlt sich immer wie eine neue Seite an - man muss sich erst vergewissern, noch auf derselben Seite zu sein
4. confusing - dieses SC will Verwirrung vermeiden. Insbesondere wenn sich das Formular, mit dem interagiert wird, under the fold befindet - also nicht im sichtbaren Bereich wenn man am Seitenanfang ist, führt das nicht nur bei Menschen mit kognitiven Einschränkungen dazu, überrascht zu sein, warum man nicht mehr beim Formular ist, ja warum das nicht einmal mehr zu sehen ist. Die Wahrscheinlichkeit, dass das Formular überraschend komplett aus dem Viewport "verschwindet" steigt mit dem Zoom-Faktor und für Menschen, die eine Vergrößerungssoftware nutzen so weit, dass es die Regel sein dürfte. Das verstärkt natürlich den unter 3 genannten Eindruck auf einer anderen Seite zu sein.
5. Falls wir uns einigen können dass hier eine neue Seite geöffnet oder der Eindruck erweckt wird, eine neue Seite wurde geöffnet, wird ein explizites Beispiel für ein Failure relevant: https://www.w3.org/WAI/WCAG21/Techniques/failures/F37

Ich finde ein Formular, bei dem mehrere Checkboxen ausgewählt werden müssen, damit das Formular Sinn macht (zum Beispiel eine Pizza mit Zutaten belegen) ist praktisch unbedienbar, wenn jede Auswahl an den Seitenanfang führt. Das führt zu ewig langen Wegen und/ oder Verwirrung für mehrere Nutzergruppen.

**Comment by detlevhfischer** [comment]
@MarcHaunschild Ich teile deine Einschätzung vollkommen, dass solch eine Fokusversetzung ein gravierendes Problem ist und ein FAIl verdient - ich habe es bislang nur bisher bei 9..2.4.3 gesehen, bin aber neugierig, was andere dazu sagen. 

Wir vermeiden ja bislang (meist?), den gleichen Fehler bei zwei verschiedenen Anforderungen anzukreiden, und ein Fokus Management Problem ist es auf jeden Fall. Soll es Deiner Meinung nach also ein doppeltes FAIL geben? In anderen Fällen bewerten wir auch den selben Mangel entweder hier oder dort - z.B. ein unverständlicher Alternativtext eines Icon-Bedienelements ist entweder ein Mangel bei 9.1.1.1a oder bei 9.2.4.6 (Beschriftung da, aber nicht aussagekräftig).

**Comment by MarcHaunschild** [comment]
Mein Fehler - habe mich zu sehr auf das konstruierte Beispiel bezogen.

Ich würde es persönlich nicht bei Fokus-Reihenfolge sehen, da der Fokus ja nicht beim Durchtabben "springt".

Es kommt zu einem neuladen und nur bei Eingabe.

Darum hätte ich es hier gesehen, nicht bei Fokus-Reihenfolge und nicht doppelt.

Eine Abgrenzung zu anderen Prüfschritten kann man gerne formulieren. Mit schien es hier nicht nötig.

Bin auch auf weitere Meinungen gespannt!

**Comment by johannesFischer84** [comment]
Hier meine Anmerkungen dazu:

1. Das Aktivieren/Deaktivieren (Ankreuzen) einer Checkbox ist eine Wertveränderung und damit eine Eingabe. Wenn nicht vorher angekündigt, wäre eine Kontextänderung unerwartet (WCAG 3.2.2). Unter Kontextveränderung wird z. B. auch eine Fokusversetzung, ein Neuladen der Seite, oder die Änderung der Scrollposition/Viewport verstanden. Meist ist das über ein _change_-Ereignis hervorgerufen. Ich würde das in den WCAG bei 3.2.2 fehlerhaft bewerten. Nicht bei 2.4.3. Die Fokusreihenfolge ist meiner Meinung nach nicht betroffen, weil die Versetzung nicht durch Navigation wie Tabulatortaste o. ä. erfolgt, sondern durch eine Wertveränderung des Formularfelds.

2. Aus meiner Sicht erklärt dies die Prüfanleitung von 9.3.2.2 auch:

> Formularelemente (Checkboxes, Radio Buttons, Auswahllisten) ausprobieren. Werden unerwartete und nicht angekündigte Kontextänderungen erzeugt?
Überprüfen, ob Inhaltsänderungen hervorgerufen werden (etwa Einblendungen zusätzlicher Formularfelder).

> Klare Kontextänderungen (changes of context) sind laut WCAG 2.1 etwa Sprünge zu anderen Seiten und das Öffnen neuer Fenster, aber auch das dynamische Laden von neuen Inhalten auf der selben Seite, wenn diese Inhalte wie eine neue Seite wirken oder Inhalte in stark veränderter Anordnung erscheinen.

Ich habe aber nichts dagegen, wenn es noch genauer erklärt würde.

**Comment by sweckenmann** [comment]
Erst einmal: Understanding 3.2.2 nimmt Bezug auf Checkboxen, Textfelder, Selects (soweit sind wir uns ja einig) :
 
> So checking a checkbox, entering text into a text field, or changing the selected option in a list control changes its setting, but activating a link or a button does not".

Wenn man in der Definition schaut, was unter "[changes of context](https://www.w3.org/WAI/WCAG21/Understanding/on-focus#dfn-changes-of-context)" gemeint ist, steht da:

> major changes in the content of the Web page that, if made without user awareness, can disorient users who are not able to view the entire page simultaneously.
> 1. Changes in context include changes of:
> 2. user agent;
> 3. viewport;
> 4. focus;
> 5. content that changes the meaning of the Web page

Somit wäre wohl jede Fokusversetzung (egal wohin) durch das Aktivieren einer Checkbox, Änderung der Option eines Selects etc., welches nicht erklärt / angekündigt wird oder welches nicht zu erwarten ist, müsste eigentlich schon ein FAIL im Sinne des Prüfschritts 9.3.2.2 sein.
(und nicht 9.2.4.3 - auch wenn ich nicht sicher bin, ob ich es bislang nicht auch eher bei 9.2.4.3 bewertet habe ;-).
Entsprechendes gilt bei "3.2.1 On Focus" bei Fokussierung.

Die Formulierung der Prüfschritt-Beschreibung (siehe oben Zitat von Johannes) scheint mir derzeit misverständlich zu sein:
> Überprüfen, ob Inhaltsänderungen hervorgerufen werden (etwa Einblendungen zusätzlicher Formularfelder).

Das Einblenden an sich ist ja nicht problematisch, nur wenn der Fokus in ein eingeblendetes Feld austomatisch versetzt wird (und das nicht erklärt / beschrieben ist). Das wird aber erst im letzten Listenpunkt der Anleitung erwähnt.

> Prüfen, ob durch Formulareingaben hervorgerufene Kontextänderungen (etwa dynamisch eingeblendete Elemente) den aktuellen Fokus versetzen.

**Comment by MarcHaunschild** [comment]
Danke @johannesFischer84, im Hinweis hatte ich das überlesen (und ich hatte mich bemüht genau zu lesen, um dann nicht so doof dazustehen wie jetzt).

Ob das nun ein 7. Punkt unter "Prüfung" sein muss (da wäre es mir persönlich aufgefallen) oder ob es reicht, das unter Hinweise stehen zu haben, ist jetzt wohl Geschmacksache.

Mir hätte ein 7. Punkt wohl geholfen und ich fände ihn deshalb sinnvoll, weil mir genau das (man aktiviert eine Checkbox und landet am Seitenanfang) immer wieder mal begegnet. Den unter 6. beschriebenen Fall habe ich dagegen noch nie gesehen.

Übrigens: ist das Einblenden von Inhalten überhaupt eine Kontextänderung (change of **context**)? Scheint mir eher eine Inhaltsänderung (change of **content**) zu sein - oder bin ich hier schon wieder auf dem Holzweg...?

**Comment by sweckenmann** [comment]
@MarcHaunschild Das Einblenden von Inhalten halte ich auch nicht für eine Kontextänderung, nur wenn z.B. bei Eingabe gleichzeitig eine Fokusversetzung stattfindet. Siehe Punkt 6 in der Prüfanleitung "Prüfen, ob durch Formulareingaben hervorgerufene Kontextänderungen (etwa dynamisch eingeblendete Elemente) den aktuellen Fokus versetzen."  Ich denke, bei der Prüfanleitung gibt es Verbesserungspotenzial (Punkt 6 müsste stärker im Zusammenhang mit Punkt 3 stehen).

**Comment by MarcHaunschild** [comment]
Nein, du hast das nicht gesagt, es steht da halt in Punkt 6  "Kontextänderungen (etwa dynamisch eingeblendete Elemente)". Der Text in Klammer soll doch eine Beispiel sein für eine Kontext-Änderung (was meiner Meinung nach falsch ist)

In Punkt 3 steht noch "Überprüfen, ob Inhaltsänderungen hervorgerufen werden (etwa Einblendungen zusätzlicher Formularfelder)" - das ist ja erst mal kein Problem, so lange der change of content später im DOM erfolgt, so dass Blinde die Änderungen bei der weiteren Lektüre bemerken.
Das wird dann in den Punkten 5 und 6 erklärt (einer davon genügt und gehört mit in Punkt 3).

Wie wäre es, den Punkt Prüfung wie folgt zusammen zu fassen:

_1. Seite im Browser (Chrome oder Firefox) aufrufen.
2. Formularelemente (Checkboxes, Radio Buttons, Auswahllisten) ausprobieren. Werden unerwartete und nicht angekündigte Kontextänderungen erzeugt (wird der Fokus zum Beispiel an eine andere Stelle wie den Seitenanfang oder sogar auf eine andere Seite gesetzt)?
3. Überprüfen, ob Inhaltsänderungen hervorgerufen werden (etwa Einblendungen zusätzlicher Formularfelder). Inhaltsänderungen müssen **nach** dem Element passieren, mit dem sie ausgelöst werden._

4 und 5 sind so in 3 mit drin und 6 scheint mir eine Dopplung von 2 - Daher entfallen diese drei Punkte.

**Comment by johannesFischer84** [comment]
> Übrigens: ist das Einblenden von Inhalten überhaupt eine Kontextänderung (change of context)? Scheint mir eher eine Inhaltsänderung (change of content) zu sein - oder bin ich hier schon wieder auf dem Holzweg...?

In der [WCAG-Kontextänderung-Definition](https://www.w3.org/TR/WCAG/#dfn-change-of-context) ist der change of content auch als change of context angegeben, allerdings mit dem Zusatz, dass die Bedeutung (meaning) der Seite verändert wird. Darunter gibt es eine NOTE, dass zusätzlich zur Inhaltsänderung auch eine Änderung des Browsers (user agent), des Viewports (Scrollposition) oder des Fokus stattfinden muss, damit es sich um eine Kontextänderung handelt. Also wäre das wohl nur eine sehr deutliche Inhaltsveränderung, was einer neuen Seite nahekommt.

**Comment by MarcHaunschild** [comment]
Vielen Dank @johannesFischer84 . Ich wusste, ich hatte das gelesen, habe aber die Stelle nciht mehr gefunden.
